src/Api/PubnubApi/EventEngine/Subscribe/Effects/ReceivingEffectHandler.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using PubnubApi.EndPoint;
+using PubnubApi.EventEngine.Common;
+using PubnubApi.EventEngine.Core;
+using PubnubApi.EventEngine.Subscribe.Context;
+using PubnubApi.EventEngine.Subscribe.Events;
+using PubnubApi.EventEngine.Subscribe.Invocations;
+using PubnubApi.EventEngine.Subscribe.Common;
+
+namespace PubnubApi.EventEngine.Subscribe.Effects
+{
+    internal class ReceivingEffectHandler:
+        Core.IEffectHandler<ReceiveMessagesInvocation>,
+        Core.IEffectHandler<ReceiveReconnectInvocation>
+    {
+        private SubscribeManager2 manager;
+        private EventQueue eventQueue;
+        
+        private Delay retryDelay = new Delay(0);
+
+        public ReceivingEffectHandler(SubscribeManager2 manager, EventQueue eventQueue)
+        {
+            this.manager = manager;
+            this.eventQueue = eventQueue;
+        }
+
+        public Task Run(ReceiveReconnectInvocation invocation)
+        {
+            if (!ReconnectionDelayUtil.shouldRetry(invocation.Policy, invocation.AttemptedRetries, invocation.MaxConnectionRetry))
+            {
+                eventQueue.Enqueue(new ReceiveReconnectGiveUpEvent() { Status = new PNStatus(PNStatusCategory.PNCancelledCategory) });
+            }
+            else
+            {
+                retryDelay = new Delay(ReconnectionDelayUtil.CalculateDelay(invocation.Policy, invocation.AttemptedRetries));
+                // Run in the background
+                retryDelay.Start().ContinueWith((_) => this.Run((ReceiveMessagesInvocation)invocation));
+            }
+
+            return Utils.EmptyTask;
+        }
+
+        public async Task Run(ReceiveMessagesInvocation invocation)
+        {
+            var response = await MakeReceiveMessagesRequest(invocation);
+
+            switch (invocation)
+            {
+                case Invocations.ReceiveReconnectInvocation reconnectInvocation when response.Item2.Error:
+                    eventQueue.Enqueue(new Events.ReceiveReconnectFailureEvent() { AttemptedRetries = reconnectInvocation.AttemptedRetries + 1, Status = response.Item2});
+                    break;
+                case Invocations.ReceiveReconnectInvocation reconnectInvocation:
+                    eventQueue.Enqueue(new Events.ReceiveReconnectSuccessEvent() { Cursor = response.Item1, Status = response.Item2 });
```

Comment:
ReceiveReconnect success will bring some messages as well I believe.
I think we are missing some info here by just accessing cursor & status isn't it.

Reply:
yeah...it is not getting messages from MakeReceiveMessagesRequest. we need to figure it out.

Reply:
I was thinking of, At least we get the payload in one property. Serialisation will be handled once we have that data!